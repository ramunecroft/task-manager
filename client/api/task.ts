import {API_URL} from "@/config";
import {type Task} from "@/server/db/schema";

export const getTasks = async () => {
  const res = await fetch(`${API_URL}/api/task`, {
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (res.headers.get("content-type") !== "application/json") {
    return;
  }
  if (res.status === 200) {
    return res.json();
  }
  return res.json();
};

export const updateTaskStatus = async (task: Task) => {
  const res = await fetch(`${API_URL}/api/task`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      title: task.title,
      description: task.description,
      ticketCode: task.ticketCode,
      status: task.status,
    }),
  });

  return res.json() as Promise<Task>;
};
