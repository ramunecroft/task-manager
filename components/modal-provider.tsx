"use client";

import {GenerateIssueModal} from "@/components/generate-issue-modal";
import {TaskModal} from "@/components/task-modal";

export const ModalProvider = ({children}: {children: React.ReactNode}) => {
  return (
    <>
      <TaskModal />
      <GenerateIssueModal />
      {children}
    </>
  );
};
