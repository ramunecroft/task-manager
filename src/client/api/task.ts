import {API_URL} from "@/app/config";
import {type DragUpdateTaskInput} from "@/server/schema";

export const getTasks = async () => {
  const res = await fetch(`${API_URL}/api/task`, {
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (res.headers.get("content-type") !== "application/json") {
    return;
  }
  if (res.status === 200) {
    return res.json();
  }
  return res.json();
};

export const updateTaskStatus = async ({
  status,
  ticketCode,
  title,
  description,
}: DragUpdateTaskInput) => {
  const res = await fetch(`${API_URL}/api/task`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      title,
      description,
      ticketCode,
      status,
    }),
  });

  return res.json();
};
