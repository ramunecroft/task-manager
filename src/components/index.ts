export * from "./search-button";
export * from "./document-modal";
