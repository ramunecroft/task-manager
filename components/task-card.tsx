"use client";

import {getTasks} from "@/client/api/task";
import {Icons} from "@/components/icons";
import {Card, CardContent, CardDescription, CardHeader} from "@/components/ui/card";
import {QUERY_KEY, priority} from "@/constants";
import {type Task} from "@/server/db/schema";
import {showTaskModalAtom, taskModalStateAtom} from "@/store/task";
import {useQuery} from "@tanstack/react-query";
import {useSetAtom} from "jotai";
import React from "react";

/**
 * title: title of task
 *
 * priority: priority of task
 */
type TaskCardProps = {
  description: string;
  ticketCode: string;
  voteCount: Task["voteCount"];
  priority: Task["priority"];
};

export const TaskCard = ({
  description,
  ticketCode,
  voteCount,
  priority,
}: TaskCardProps) => {
  const {
    data: taskList,
    isLoading,
    isError,
  } = useQuery<Task[]>({
    queryKey: [QUERY_KEY.tasklist],
    queryFn: () => getTasks(),
  });

  const setShowTaskModal = useSetAtom(showTaskModalAtom);
  const setTaskModal = useSetAtom(taskModalStateAtom);

  if (!taskList) return null;

  const handleDragStart = (e: React.DragEvent<HTMLDivElement>, ticketCode: string) => {
    e.dataTransfer.setData("text/plain", ticketCode);
  };

  const onClick = () => {
    setShowTaskModal(true);
    const findTask = taskList?.find(el => el.ticketCode === ticketCode);
    if (!findTask) return;
    setTaskModal(findTask);
  };

  return (
    <Card
      role="task-card"
      onClick={() => onClick()}
      onDragStart={e => handleDragStart(e, ticketCode)}
      className="w-full flex-1 hover:cursor-pointer hover:shadow-lg"
      draggable>
      <CardHeader>
        <CardDescription className="line-clamp-4 break-words dark:text-white">
          {description}
        </CardDescription>
      </CardHeader>
      <CardContent className="my-1 flex flex-1 flex-row items-center justify-between pb-4">
        <div className="flex items-center justify-center space-x-2">
          <Icons.story />
          <div className="py-1 text-sm font-semibold text-green-700">{ticketCode}</div>
        </div>
        <div className="flex items-center justify-center space-x-2">
          <div className="flex h-6 w-6 items-center justify-center rounded-full bg-gray-200">
            <span className="text-sm font-semibold  text-gray-700">{voteCount}</span>
          </div>
          <PriorityIcon level={priority} />
          <Icons.user />
        </div>
      </CardContent>
    </Card>
  );
};

export const PriorityIcon = ({level}: {level: Task["priority"]}) => {
  if (level === priority.lower) {
    return <Icons.lower data-testid={123} />;
  }
  if (level === priority.low) {
    return <Icons.low />;
  }
  if (level === priority.medium) {
    return <Icons.medium data-testid={125} />;
  }
  if (level === priority.high) {
    return <Icons.high />;
  }
  if (level === priority.highest) {
    return <Icons.highest />;
  }
};
