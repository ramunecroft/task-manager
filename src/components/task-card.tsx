"use client";

import {Priority} from "@/constants";
import {Icons} from "@/components/icons";
import {
  Card,
  CardHeader,
  CardTitle,
  CardContent,
  CardDescription,
} from "@/components/ui/card";
import {type Task} from "@/server/schema";
import React from "react";

/**
 * title: title of task
 *
 * priority: priority of task
 */
type TaskCardProps = {
  description: string;
  ticketCode: string;
  voteCount: Task["voteCount"];
  priority: Task["priority"];
};

export const TaskCard = ({
  description,
  ticketCode,
  voteCount = 0,
  priority,
}: TaskCardProps) => {
  const handleDragStart = (e: React.DragEvent<HTMLDivElement>, ticketCode: string) => {
    e.dataTransfer.setData("text/plain", ticketCode);
  };

  return (
    <Card
      onDragStart={e => handleDragStart(e, ticketCode)}
      className="w-full flex-1 hover:cursor-pointer hover:shadow-lg"
      draggable>
      <CardHeader>
        <CardDescription className="line-clamp-4 break-words dark:text-white">
          {description}
        </CardDescription>
      </CardHeader>
      <CardContent className="my-1 flex flex-1 flex-row items-center justify-between pb-4">
        <div className="flex items-center justify-center space-x-2">
          <Icons.story />
          <div className="py-1 text-sm font-semibold text-green-700">{ticketCode}</div>
        </div>
        <div className="flex items-center justify-center space-x-2">
          <div className="flex h-6 w-6 items-center justify-center rounded-full bg-gray-200">
            <span className="text-sm font-semibold  text-gray-700">{voteCount}</span>
          </div>
          <PriorityIcon level={priority} />
          <Icons.user />
        </div>
      </CardContent>
    </Card>
  );
};

const PriorityIcon = ({level}: {level: TaskCardProps["priority"]}) => {
  if (level === Priority.Lower) {
    return <Icons.lower />;
  }
  if (level === Priority.Low) {
    return <Icons.low />;
  }
  if (level === Priority.Medium) {
    return <Icons.medium />;
  }
  if (level === Priority.High) {
    return <Icons.high />;
  }
  if (level === Priority.Highest) {
    return <Icons.highest />;
  }
};
