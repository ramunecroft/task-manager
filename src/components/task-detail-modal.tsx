"use client";

import {updateTaskStatus} from "@/client/api/task";
import {Icons} from "@/components/icons";
import {PriorityIcon} from "@/components/task-card";
import {
  Collapsible,
  CollapsibleContent,
  CollapsibleTrigger,
} from "@/components/ui/collapsible";
import {Dialog, DialogContent} from "@/components/ui/dialog";
import {
  Select,
  SelectContent,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from "@/components/ui/select";
import {type Task} from "@/server/schema";
import {
  mutateAtom,
  selectedTaskAtom,
  showTaskModalAtom,
  taskListAtom,
} from "@/store/task";
import {useAtom, useAtomValue, useSetAtom} from "jotai";
import {UserIcon} from "lucide-react";

export const TaskDetailModal = () => {
  const [showTaskModal, setShowTaskModal] = useAtom(showTaskModalAtom);
  const taskList = useAtomValue(taskListAtom);
  const mutate = useSetAtom(mutateAtom);
  const selectedTask = useAtomValue(selectedTaskAtom);
  if (!selectedTask) return;

  const onStatusChange = async (status: Task["status"]) => {
    await updateTaskStatus({
      status,
      ticketCode: selectedTask.ticketCode,
    });

    const updated = taskList.map(task => {
      if (task.ticketCode === selectedTask.ticketCode) {
        task.status = status;
      }
      return task;
    });
    mutate(updated);
  };

  return (
    <Dialog open={showTaskModal} onOpenChange={setShowTaskModal}>
      <DialogContent
        role={"task-modal"}
        className="mx-auto my-8 flex h-3/4 max-w-md flex-col items-start rounded-lg bg-white p-4 shadow-lg sm:max-w-xl md:max-w-6xl">
        <div className="flex flex-col gap-y-2">
          <div id="tool-tip w-full">
            <div className="flex w-full flex-grow flex-row items-center justify-between">
              <div className="flex items-center justify-center space-x-2">
                <PriorityIcon level={selectedTask.priority} />
                <p>{selectedTask.ticketCode}</p>
              </div>
              <div className="mr-6 flex gap-x-2">
                <Icons.eye />
                <p>{selectedTask.viewCount}</p>
              </div>
            </div>
          </div>
          <div id="content" className="flex flex-row gap-x-8">
            <div id="left-content" className="basis-4/6 space-y-2">
              <div className="flex flex-grow items-start justify-start">
                <p className="text-3xl font-bold">{selectedTask.title}</p>
              </div>
              <div className="flex flex-grow justify-start">
                <UserIcon />
                <UserIcon />
                <UserIcon />
                <UserIcon />
                <UserIcon />
                <UserIcon />
              </div>
              <p className="text-gray-400">Description</p>
              <div className="flex flex-col bg-gray-100 p-1">
                {selectedTask.description}
              </div>
            </div>
            <div id="right-content" className="flex basis-2/6 flex-col">
              <Select onValueChange={(e: Task["status"]) => onStatusChange(e)}>
                <SelectTrigger className="w-[180px] bg-gray-100 font-extrabold">
                  <SelectValue placeholder={selectedTask.status.split("_").join(" ")} />
                </SelectTrigger>
                <SelectContent className="font-extrabold">
                  <SelectItem value="TO_DO">TO DO</SelectItem>
                  <SelectItem value="IN_PROGRESS">IN PROGRESS</SelectItem>
                  <SelectItem value="IN_REVIEW">IN REVIEW</SelectItem>
                  <SelectItem value="DONE">DONE</SelectItem>
                </SelectContent>
              </Select>
              <div id="task-detail">
                <Collapsible>
                  <CollapsibleTrigger>Can I use this in my project?</CollapsibleTrigger>
                  <CollapsibleContent>
                    Yes. Free to use for personal and commercial projects. No
                    attribution required.
                  </CollapsibleContent>
                </Collapsible>
              </div>
            </div>
          </div>
        </div>
      </DialogContent>
    </Dialog>
  );
};
