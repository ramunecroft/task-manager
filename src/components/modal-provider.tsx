"use client";

import {TaskDetailModal} from "@/components/task-detail-modal";

export const ModalProvider = ({children}: {children: React.ReactNode}) => {
  return (
    <>
      <TaskDetailModal />
      {children}
    </>
  );
};
