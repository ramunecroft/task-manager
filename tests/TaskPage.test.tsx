import TaskPage from "@/app/task/page";
import {ModalProvider} from "@/components/modal-provider";
import {TaskCard} from "@/components/task-card";
import {TaskDetailModal} from "@/components/task-detail-modal";
import {showTaskModalAtom} from "@/store/task";
import "@testing-library/jest-dom";
import {fireEvent, render, renderHook, screen} from "@testing-library/react";
import {useAtomValue} from "jotai";

describe("Page render test", () => {
  it("renders a heading", () => {
    const {container} = render(<TaskPage />);
    expect(container).toMatchSnapshot();

    const testDescription = "hello";
    const testTicketCode = "T3";
    const testVoteCount = 5;

    render(
      <TaskCard
        description={testDescription}
        priority={"low"}
        ticketCode={testTicketCode}
        voteCount={testVoteCount}
      />
    );

    const descriptionElement = screen.getByText(testDescription);
    const ticketCodeElement = screen.getByText(testTicketCode);
    const voteCountElement = screen.getByText(testVoteCount);
    expect(descriptionElement).toBeInTheDocument();
    expect(ticketCodeElement).toBeInTheDocument();
    expect(voteCountElement).toBeInTheDocument();
  });
});

describe("TaskCard on Click Test", () => {
  it("button", () => {
    const showTaskModal = renderHook(() => useAtomValue(showTaskModalAtom));
    const testDescription = "hello";
    const testTicketCode = "T3";
    const testVoteCount = 5;
    const {getByRole} = render(
      <TaskCard
        description={testDescription}
        priority={"low"}
        ticketCode={testTicketCode}
        voteCount={testVoteCount}
      />
    );

    fireEvent.click(getByRole("task-card"));

    expect(showTaskModal.result.current).toBe(true);
  });
});
