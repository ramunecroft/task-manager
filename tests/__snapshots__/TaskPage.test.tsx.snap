// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page render test renders a heading 1`] = `<div />`;

exports[`Page renders a heading 1`] = `<div />`;
